src/btree/bt_split.c

```diff
@@ -1431,20 +1431,31 @@ __split_multi_inmem(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *multi, WT
          * size of the page is correct.
          */
         if (supd->onpage_upd != NULL && !F_ISSET(S2C(session), WT_CONN_IN_MEMORY)) {
+            tmp = supd->onpage_tombstone != NULL ? supd->onpage_tombstone : supd->onpage_upd;
+
             /*
              * We have decided to restore this update chain so it must have newer updates than the
              * onpage value on it.
              */
-            WT_ASSERT(session, upd != supd->onpage_upd);
+            WT_ASSERT(session, upd != tmp);
+            WT_ASSERT(session, F_ISSET(tmp, WT_UPDATE_DS));
+
             /*
              * Move the pointer to the position before the onpage value and truncate all the updates
              * starting from the onpage value.
              */
-            for (prev_onpage = upd;
-                 prev_onpage->next != NULL && prev_onpage->next != supd->onpage_upd;
+            for (prev_onpage = upd; prev_onpage->next != NULL && prev_onpage->next != tmp;
                  prev_onpage = prev_onpage->next)
                 ;
-            WT_ASSERT(session, prev_onpage->next == supd->onpage_upd);
+            WT_ASSERT(session, prev_onpage->next == tmp);
+#ifdef HAVE_DIAGNOSTIC
+            /*
+             * Relies on the assumption that the tombstone is the only thing before the onpage
```

Comment:
Better phrasing:
During update restore eviction we remove anything older than the on-page update including the on-page update. However it is possible a tombstone is also written as the stop time of the on-page value. To handle this we need to remove the tombstone too from the update chain.

This assertion checks that there aren't any unexpected updates between that tombstone and the subsequent value which both make up the on-page value.

Reply:
Thanks, fixed in 9169207